src/monitor/node_active_protocol.c

```diff
@@ -892,7 +972,11 @@ remove_node(PG_FUNCTION_ARGS)
 	int32 nodePort = PG_GETARG_INT32(1);
 
 	AutoFailoverNode *currentNode = NULL;
-	AutoFailoverNode *otherNode = NULL;
+	bool currentNodeIsPrimary = false;
```

Comment:
I think  we still have some problems around dropping the primary node from the cluster. This is not blocking my testing, but we should fix it @DimCitus

```
 pg_autoctl show state --pgdata /Users/onderkalaci/Documents/pg_auto_failover/monitor/
                                    Name |   Port | Group |  Node |     Current State |    Assigned State | Priority | Quorum
-----------------------------------------+--------+-------+-------+-------------------+-------------------+----------+-------
cusred034c9409-vl2.network.microsoft.com |  10001 |     0 |     1 |           primary |           primary |       50 |      t
cusred034c9409-vl2.network.microsoft.com |  10002 |     0 |     2 |         secondary |         secondary |       50 |      t
cusred034c9409-vl2.network.microsoft.com |  10003 |     0 |     3 |         secondary |         secondary |       50 |      t

-- the first time to call `--destroy` fails. which is slightly confusing as well
pg_autoctl drop node --pgdata /Users/onderkalaci/Documents/pg_auto_failover/worker1a --destroy
10:11:12 94286 INFO  An instance of pg_autoctl is running with PID 93697, stopping it.
10:11:12 94286 INFO  Removing local node from the pg_auto_failover monitor.
10:11:12 94286 ERROR Monitor ERROR:  ProceedGroupState couldn't find the primary node in formation "default", group 0
10:11:12 94286 ERROR Monitor DETAIL:  activeNode is cusred034c9409-vl2.network.microsoft.com:10002 in state secondary
10:11:12 94286 ERROR SQL query: SELECT pgautofailover.remove_node($1, $2)
10:11:12 94286 ERROR SQL params: 'cusred034c9409-vl2.network.microsoft.com', '10001'
10:11:12 94286 ERROR Failed to remove node cusred034c9409-vl2.network.microsoft.com:10001 from the monitor
10:11:12 94286 INFO  Removing local node state file: "/Users/onderkalaci/.local/share/pg_autoctl/Users/onderkalaci/Documents/pg_auto_failover/worker1a/pg_autoctl.state"
10:11:12 94286 INFO  Removing local node init state file: "/Users/onderkalaci/.local/share/pg_autoctl/Users/onderkalaci/Documents/pg_auto_failover/worker1a/pg_autoctl.init"
10:11:12 94286 FATAL Failed to remove local node from the pg_auto_failover monitor, see above for details


-- the second time, it works
pg_autoctl drop node --pgdata /Users/onderkalaci/Documents/pg_auto_failover/worker1a --destroy
10:11:17 94320 WARN  Skipping node removal from the monitor: state file "/Users/onderkalaci/.local/share/pg_autoctl/Users/onderkalaci/Documents/pg_auto_failover/worker1a/pg_autoctl.state" does not exist
10:11:17 94320 INFO  Stopping PostgreSQL at "/Users/onderkalaci/Documents/pg_auto_failover/worker1a"
10:11:17 94320 INFO  /Users/onderkalaci/Documents/citus_code/pgenv/pgsql/bin//pg_ctl --pgdata /Users/onderkalaci/Documents/pg_auto_failover/worker1a --wait stop --mode fast
10:11:17 94320 INFO  /Users/onderkalaci/Documents/citus_code/pgenv/pgsql/bin//pg_ctl status -D /Users/onderkalaci/Documents/pg_auto_failover/worker1a [3]
10:11:17 94320 INFO  pg_ctl: no server running
10:11:17 94320 INFO  pg_ctl stop failed, but PostgreSQL is not running anyway
10:11:17 94320 INFO  Removing "/Users/onderkalaci/Documents/pg_auto_failover/worker1a"
10:11:17 94320 INFO  Removing "/Users/onderkalaci/.config/pg_autoctl/Users/onderkalaci/Documents/pg_auto_failover/worker1a/pg_autoctl.cfg"
cusred034c9409-vl2:pg_auto_failover onderkalaci$ pg_autoctl drop node --pgdata /Users/onderkalaci/Documents/pg_auto_failover/worker1a --destroy


-- but the primary still  shows up even after minutes has pass
pg_autoctl show state --pgdata /Users/onderkalaci/Documents/pg_auto_failover/monitor/
                                    Name |   Port | Group |  Node |     Current State |    Assigned State | Priority | Quorum
-----------------------------------------+--------+-------+-------+-------------------+-------------------+----------+-------
cusred034c9409-vl2.network.microsoft.com |  10001 |     0 |     1 |           primary |           demoted |       50 |      t
cusred034c9409-vl2.network.microsoft.com |  10002 |     0 |     2 |           primary |           primary |       50 |      t
cusred034c9409-vl2.network.microsoft.com |  10003 |     0 |     3 |         secondary |         secondary |       50 |      t


-- monitor sees it as unhealthy as the node is not running already

 SELECT nodeid, health FROM pgautofailover.node;
 nodeid | health 
--------+--------
      3 |      1
      1 |      0
      2 |      1
(3 rows)

-- the removed node still shows up in sync. standby names
pg_autoctl show synchronous_standby_names --pgdata /Users/onderkalaci/Documents/pg_auto_failover/monitor/
ANY 1 (pgautofailover_standby_3, pgautofailover_standby_1)

-- but at least, writes can still continue
pg_autoctl get formation   number-sync-standbys --pgdata /Users/onderkalaci/Documents/pg_auto_failover/worker1b
1

Reply:
I see, thanks. I think when we're removing the primary node we should do that in three steps:

  1. set the node to DRAINING and the other nodes to REPORT_LSN, 
  2. call `ProceedGroupState` to have the monitor FSM make some progress,
  3. and only then remove the primary

On the other hand, when we're removing any other node than the primary, we should first remove the node from the monitor and then call `ProceedGroupState` on the primary, as we used to when we had only two nodes per group.